classes/order/Order.php

```diff
@@ -1827,13 +1827,19 @@ public function getOrderDetailList()
 
     /**
      * Generate a unique reference for orders generated with the same cart id
-     * This references, is useful for check payment.
+     * This reference is the primary order identifier for public use.
+     *
+     * Modules can return their own reference.
      *
      * @return string
      */
     public static function generateReference()
     {
-        return strtoupper(Tools::passwdGen(9, 'NO_NUMERIC'));
+        $reference = Hook::exec('actionGenerateDocumentReference', [
+            'type' => 'order',
+        ]);
+
+        return !empty($reference) ? $reference : strtoupper(Tools::passwdGen(9, 'NO_NUMERIC'));
```

Comment:
Litte improvement suggestion
```suggestion
        return $reference ?: strtoupper(Tools::passwdGen(9, 'NO_NUMERIC'));
```

Reply:
🤔 I'm not sure it will improve something, the ternary operator should be used with an expression (in this case the empty() function), your suggestion it's like doing:
`if($reference)`
in case $_reference_ is not populated (i.e. it's undefined) it will generate a _notice_ 

wdyt?

Reply:
`$reference` will always be null or a string , I'd prefer `if ((string) $reference !== '') {}`

That's a personal taste (I really don't like `empty` TBH). You can leave it as it is :)

Reply:
@marsaldev @FabienPapet Guys, I will leave it as it is for now, we can improve later, if you make a deal. 😄😄

Reply:
@Hlavtox 
ahah! no worries, I'm always curious about a choice instead of another, and always interested to understand the reason of that choice 😊